aiohttp/http_writer.py

```diff
@@ -125,7 +128,18 @@ def drain(self):
           yield from w.drain()
         """
         if self._protocol.transport is not None:
-            yield from self._protocol._drain_helper()
+            if self._is_drain:
+                fut = asyncio.Future(loop=self._loop)
+                self._drain_waiters.append(fut)
+                yield from fut
+            self._is_drain = True
+            try:
+                yield from self._protocol._drain_helper()
+            finally:
+                self._is_drain = False
+                if self._drain_waiters:
+                    fut = self._drain_waiters.popleft()
+                    fut.set_result(None)
```

Comment:
The same "deadlock" issue (a,b) can happen here: if future on line 134 is cancelled then other task that
sets result for next waiter will do nothing and all other futures will be in `_drain_waiters` forever.

a: bugs.python.org/issue27585
b: https://github.com/python/asyncio/pull/393 and propbably https://github.com/python/cpython/pull/1031